confgenerator/feature.go

```diff
@@ -0,0 +1,273 @@
+		tempTrackedFeatures, err = trackedMappedComponents("metrics", "receivers", uc.Metrics.Receivers)
+		if err != nil {
+			return nil, err
+		}
+		allFeatures = append(allFeatures, tempTrackedFeatures...)
+
+		tempTrackedFeatures, err = trackedMappedComponents("metrics", "processors", uc.Metrics.Processors)
+		if err != nil {
+			return nil, err
+		}
+		allFeatures = append(allFeatures, tempTrackedFeatures...)
+	}
+
+	if uc.HasLogging() {
+		tempTrackedFeatures, err = trackedMappedComponents("logging", "receivers", uc.Logging.Receivers)
+		if err != nil {
+			return nil, err
+		}
+		allFeatures = append(allFeatures, tempTrackedFeatures...)
+
+		tempTrackedFeatures, err = trackedMappedComponents("logging", "processors", uc.Logging.Processors)
+		if err != nil {
+			return nil, err
+		}
+		allFeatures = append(allFeatures, tempTrackedFeatures...)
+	}
+	return allFeatures, nil
+}
+
+func getSortedKeys[C Component](m map[string]C) []string {
+	keys := make([]string, 0, len(m))
+	for k := range m {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+func trackedMappedComponents[C Component](module string, kind string, m map[string]C) ([]Feature, error) {
+	if m == nil {
+		return nil, nil
+	}
+	var features []Feature
+	index := 0
+	for _, k := range getSortedKeys(m) {
+		c := m[k]
+		feature := Feature{
+			Module: module,
+			Kind:   kind,
+			Type:   c.Type(),
+			Key:    []string{fmt.Sprintf("[%d]", index)},
+		}
+		index++
+		trackedFeatures, err := trackingFeatures(reflect.ValueOf(c), metadata{}, feature)
+		if err != nil {
+			return nil, err
+		}
+		features = append(features, trackedFeatures...)
+	}
+
+	return features, nil
+}
+
+func trackingFeatures(c reflect.Value, m metadata, feature Feature) ([]Feature, error) {
+	if m.IsExcluded {
+		return nil, nil
+	}
+	t := c.Type()
+
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	v := reflect.Indirect(c)
+	if v.Kind() == reflect.Invalid || v.IsZero() {
+		return nil, nil
+	}
+
+	var features []Feature
+
+	switch kind := t.Kind(); {
+	case kind == reflect.Struct:
+		// If struct has tracking it must have a value
+		if m.HasTracking && !m.HasOverride {
+			return nil, ErrTrackingOverrideStruct
+		}
+
+		if m.YamlTag != "" {
+			feature.Key = append(feature.Key, m.YamlTag)
+		}
+
+		if m.HasTracking {
+			// If struct is inline there is no associated name for key generation
+			// By default inline structs of a tracked field are also tracked
+			if m.IsInline {
+				return nil, ErrTrackingInlineStruct
+			} else {
+				// For structs that are in a Component. An extra metric is added with
+				// the value being the override value from the yaml tag
+				ftr := feature
+				ftr.Value = m.OverrideValue
+				features = append(features, ftr)
+			}
+		}
+
+		// Iterate over all available fields and read the tag value
+		for i := 0; i < t.NumField(); i++ {
+			// Get the field, returns https://golang.org/pkg/reflect/#StructField
+			field := t.Field(i)
+			if field.Name == "Type" {
```

Comment:
We should document here what this is. Basically the `Type` field name is part of the `ConfigComponent` definition. And all user visible component inlines that component. That means this field can help us assert that a certain component is enabled.